sklearn/experimental/_column_transformer.py

```diff
@@ -0,0 +1,454 @@
+            contained subobjects that are estimators.
+
+        Returns
+        -------
+        params : mapping of string to any
+            Parameter names mapped to their values.
+        """
+        return self._get_params('_transformers', deep=deep)
+
+    def set_params(self, **kwargs):
+        """Set the parameters of this estimator.
+
+        Valid parameter keys can be listed with ``get_params()``.
+
+        Returns
+        -------
+        self
+        """
+        self._set_params('_transformers', **kwargs)
+        return self
+
+    def _iter(self, X=None, skip_none=True, fitted=False):
+        """Generate (name, trans, column, weight) tuples
+        """
+        if fitted:
+            transformers = self.transformers_
+        else:
+            transformers = self.transformers
+        get_weight = (self.transformer_weights or {}).get
+        return ((name, trans, _get_column(X, column) if X is not None else X,
+                 get_weight(name))
+                for name, trans, column in transformers
+                if not skip_none or trans is not None)
+
+    def _validate_transformers(self):
+        names, transformers, _, _ = zip(*self._iter(skip_none=False))
+
+        # validate names
+        self._validate_names(names)
+
+        # validate estimators
+        for t in transformers:
+            if t is None:
+                continue
+            if (not (hasattr(t, "fit") or hasattr(t, "fit_transform")) or not
+                    hasattr(t, "transform")):
+                raise TypeError("All estimators should implement fit and "
+                                "transform. '%s' (type %s) doesn't" %
+                                (t, type(t)))
+
+    @ property
+    def named_transformers_(self):
+        """
+        Read-only attribute to access any transformer by given name.
+        Keys are transformer names and values are the fitted transformer
+        objects.
+
+        """
+        # Use Bunch object to improve autocomplete
+        return Bunch(**dict([(name, trans) for name, trans, _
+                             in self.transformers_]))
+
+    def get_feature_names(self):
+        """Get feature names from all transformers.
+
+        Returns
+        -------
+        feature_names : list of strings
+            Names of the features produced by transform.
+        """
+        check_is_fitted(self, 'transformers_')
+        feature_names = []
+        for name, trans, _, _ in self._iter(fitted=True):
+            if not hasattr(trans, 'get_feature_names'):
+                raise AttributeError("Transformer %s (type %s) does not "
+                                     "provide get_feature_names."
+                                     % (str(name), type(trans).__name__))
+            feature_names.extend([name + "__" + f for f in
+                                  trans.get_feature_names()])
+        return feature_names
+
+    def _update_fitted_transformers(self, transformers):
```

Comment:
Perhaps consider combining fit and fit transform into one function and roll this in. I.e.
```
def fit(self, X, y=None):
    return self._fit(X, y, transform=False)

def _fit(self, X, y, transform):
    if transform:
        func = _fit_transform_one
    else:
        func = _fit_one_transformer
    ...
    if transform:
        Xs, transformers = result
    else:
        transformers = result
    # TODO insert _update... here
    if not transform:
        return self
    # TODO calculate return value in transform case
```
I just think it will be easier to see the shared logic.